### PR TITLE
perf(Role): linear speed position getter

### LINKED ISSUE #3925: Sorting members by highest role position takes extremely long in large servers (10k+ members)
**Please describe the problem you are having in as much detail as possible:**

I made a function for the bot that finds members by their username, useful for moderation commands and such. Members with higher roles are prioritized, so the member list is sorted by highest role position to lowest. This process was near instant in v11, but ever since updating to v12 it has been taking significantly longer, sometimes up to **three minutes** in servers with over 10k members. The bot freezes during this process and no event handlers fire.

**Include a reproducible code sample here, if possible:**

```js
let startTime = Date.now()
let sortedMembers = message.guild.members.cache.sort(function(x, y){return y.roles.highest.position - x.roles.highest.position})
console.log(Date.now() - startTime)  // 180117 (180 seconds)
console.log(sortedMembers.size)      // 13210
```

**Further details:**

- discord.js version: 12.0.2
- Node.js version: 12.6.1
- Operating system: Windows 10
- Priority this issue should have – please be realistic and elaborate if possible: Somewhat low

- [ ] I have also tested the issue on latest master, commit hash:

### PR DESCRIPTION
**Pull request by kyranet**
**Please describe the changes this PR makes and why it should be merged:**

Alternative to #9490, fixes #3925.

| Iterations |      Old |      New | Difference |
| :--------: | :------: | :------: | :--------: |
|     10_000 | 631.24ms | 174.44ms |     27.63% |
|    100_000 |    8.01s |    1.89s |     23.59% |

Basically skips the `sorted` call, which is an algorithm with an O(n * log(n)) cost, as well as the indexOf, which is a O(n) call. With this, not only it's not longer allocating a Collection nor an Array (making the only allocation be the iterator used in reduce), but the performance also goes down to a single O(n).

### <samp>🤖 Generated by Copilot at 8b28bc4</samp>

> _`position` getter_
> _improved with reduce function_
> _snowflakes in winter_

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by jaw0r3k** [comment]
Can you make the same for channels?

**Comment by kyranet** [comment]
Yes, in a follow-up PR, I wanted to push this one first to get reviews in case the code can be improved any further.

**Comment by jaw0r3k** [comment]
> Yes, in a follow-up PR, I wanted to push this one first to get reviews in case the code can be improved any further.

Idk if using simply for loop will be a little bit faster, as not calling the function everytime
Except that it looks good

File: packages/discord.js/src/structures/Role.js [hunk 1]
```diff
@@ -188,8 +188,14 @@ class Role extends Base {
    * @readonly
    */
   get position() {
-    const sorted = this.guild._sortedRoles();
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Role.js:191]: Doesn't this mean we no longer need _sortedRoles? 👀 🧹

Comment (jaw0r3k) [reply packages/discord.js/src/structures/Role.js:191]: It is used in setting position

Comment (kyranet) [reply packages/discord.js/src/structures/Role.js:191]: We still do:

https://github.com/discordjs/discord.js/blob/b327f4925ff557e0aa8273d7f017aa616226ca06/packages/discord.js/src/managers/RoleManager.js#L252

**Review by iCrawl — APPROVED** [review]
🐎

**Review by SpaceEEC — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**